File: ToyLanguage5/src/Model/closeRFile.java
package Model;

import javafx.util.Pair;

import java.io.BufferedReader;
import java.io.IOException;

public class closeRFile implements IStmt {
    private Exp exp_file_id;

    public closeRFile(Exp exp_file_id) {
        this.exp_file_id = exp_file_id;
    }

    @Override
    public PrgState execute(PrgState state) throws MyException, IOException {
        MyIDictionary<String,Integer> symTbl = state.getSymTable();
        MyIFileDictionary<Integer, Pair<String, BufferedReader>> fileTable = state.getFileTable();
        MyIHeap<Integer,Integer> hp = state.getHeap();
        try
        {
            int val = exp_file_id.eval(symTbl,hp);
            Pair<String,BufferedReader> p = fileTable.lookup(val);
            BufferedReader br = p.getValue();
            br.close();
            fileTable.remove(val); //not Sure
        }
        catch (MyException ex) {
            throw new MyException("" + ex);
        }

        return  state;
    }

    @Override
    public String toString() {
        return "closeRFile(" +exp_file_id +')';
    }
}

File: ToyLanguage7/ToyLanguage7/CloseRFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace ToyLanguage7
{
    class CloseRFile : IStmt
    {
        private Exp exp_file_id;

        public CloseRFile(Exp exp_file_id)
        {
            this.exp_file_id = exp_file_id;
        }


        public PrgState execute(PrgState state)
        {
            MyIDictionary<string, int> symTbl = state.getSymTable();
            MyIFileDictionary<int, KeyValuePair<string, StreamReader>> fileTable = state.getFileTable();
            try
            {
                int val = exp_file_id.eval(symTbl);
                KeyValuePair<string, StreamReader> p = fileTable.lookup(val);
                StreamReader br = p.Value;
                br.Close();
                fileTable.remove(val); //not Sure
            }
            catch (MyException ex)
            {
                throw new MyException("" + ex);
            }

            return state;
        }

        public override string ToString()
        {
            return "closeRFile(" + exp_file_id + ')';
        }
    }
}
